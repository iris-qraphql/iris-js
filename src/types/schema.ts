import type { DirectiveNode, ParseOptions, Source } from 'graphql';
import { prop, uniqBy } from 'ramda';

import { typeCheckASTValue } from '../validation/typeCheckASTValue';
import { validateSDL } from '../validation/validate';

import type { IrisError } from '../error';
import { irisNodeError } from '../error';
import { parse } from '../parsing';
import type { TypeMap } from '../utils/collectTypeMap';
import { collectTypeMap } from '../utils/collectTypeMap';
import { inspect } from '../utils/legacy';
import type { IrisMaybe, Maybe } from '../utils/type-level';
import { notNill } from '../utils/type-level';

import type {
  ArgumentDefinitionNode,
  DirectiveDefinitionNode,
  FieldDefinitionNode,
  NamedTypeNode,
  Role,
  TypeDefinitionNode,
  TypeNode,
  VariantDefinitionNode,
} from './ast';
import type {
  IrisArgument,
  IrisField,
  IrisType,
  IrisVariant,
} from './definition';
import { IrisScalars, IrisTypeDefinition, IrisTypeRef } from './definition';
import {
  GraphQLDeprecatedDirective,
  GraphQLDirective,
  specifiedDirectives,
} from './directives';
import { IrisKind } from './kinds';

class IrisSchema {
  description: Maybe<string>;

  readonly query?: IrisTypeDefinition<'resolver'>;
  readonly mutation?: IrisTypeDefinition<'resolver'>;
  readonly subscription?: IrisTypeDefinition<'resolver'>;
  readonly directives: ReadonlyArray<GraphQLDirective>;
  readonly typeMap: TypeMap;

  // Used as a cache for validateSchema().
  __validationErrors: Maybe<ReadonlyArray<IrisError>>;

  constructor({
    description,
    query,
    mutation,
    subscription,
    types = [],
    directives = [],
  }: Readonly<IrisSchemaConfig>) {
    this.description = description;
    this.query = query;
    this.mutation = mutation;
    this.subscription = subscription;
    this.directives = uniqBy(prop('name'), [
      ...directives,
      ...specifiedDirectives,
    ]);

    this.typeMap = collectTypeMap(
      [query, mutation, subscription, ...types].filter(notNill),
      this.directives,
    );
  }

  get [Symbol.toStringTag]() {
    return 'IrisSchema';
  }

  getType = (name: string): IrisMaybe<IrisTypeDefinition> => this.typeMap[name];

  getDirective = (name: string): Maybe<GraphQLDirective> =>
    this.directives.find((directive) => directive.name === name);
}

export interface IrisSchemaConfig {
  description?: Maybe<string>;
  query?: IrisTypeDefinition<'resolver'>;
  mutation?: IrisTypeDefinition<'resolver'>;
  subscription?: IrisTypeDefinition<'resolver'>;
  types?: ReadonlyArray<IrisTypeDefinition>;
  directives?: ReadonlyArray<GraphQLDirective>;
}

export function buildSchema(
  source: string | Source,
  options?: ParseOptions,
): IrisSchema {
  const documentAST = parse(source, { noLocation: options?.noLocation });
  const errors = validateSDL(documentAST);

  if (errors.length !== 0) {
    throw new Error(errors.map((error) => error.message).join('\n\n'));
  }

  const directiveDefs: Array<DirectiveDefinitionNode> = [];
  const typeMap: Record<string, IrisTypeDefinition> = {};

  function lookupType<R extends Role>(
    node: NamedTypeNode | VariantDefinitionNode<R>,
  ): IrisTypeDefinition<R> {
    const name = node.name.value;
    const type = IrisScalars[name] ?? typeMap[name];

    if (type === undefined) {
      throw new Error(`Unknown type: "${name}".`);
    }

    return type as IrisTypeDefinition<R>;
  }

  function getWrappedType(node: TypeNode): IrisType {
    if (node.kind === IrisKind.LIST_TYPE) {
      return new IrisTypeRef('LIST', getWrappedType(node.type));
    }
    if (node.kind === IrisKind.MAYBE_TYPE) {
      return new IrisTypeRef('MAYBE', getWrappedType(node.type));
    }
    return lookupType(node);
  }

  function buildDirective(node: DirectiveDefinitionNode): GraphQLDirective {
    return new GraphQLDirective({
      name: node.name.value,
      description: node.description?.value,
      locations: node.locations.map(({ value }) => value as any),
      isRepeatable: node.repeatable,
      args: node.arguments?.map(buildArgument),
      astNode: node,
    });
  }

  function buildArgument(astNode: ArgumentDefinitionNode): IrisArgument {
    const type: any = getWrappedType(astNode.type);
    const name = astNode.name.value;
    return {
      name,
      type,
      description: astNode.description?.value,
      defaultValue: typeCheckASTValue(astNode.defaultValue, type),
      deprecationReason: getDeprecationReason(astNode),
      astNode,
    };
  }

  const buildField = <R extends Role>(
    field: FieldDefinitionNode<R>,
  ): IrisField<R> =>
    ({
      name: field.name.value,
      type: getWrappedType(field.type),
      description: field.description?.value,
      deprecationReason: getDeprecationReason(field),
      astNode: field,
      args: field.arguments?.map(buildArgument),
      toJSON: () => field.name.value,
    } as IrisField<R>);

  const buildVariant = <R extends Role>(
    astNode: VariantDefinitionNode<R>,
  ): IrisVariant<R> => {
    const name = astNode.name.value;
    const description = astNode.description?.value;
    const deprecationReason = getDeprecationReason(astNode);

    if (!astNode.fields) {
      return {
        name,
        description,
        deprecationReason,
        astNode,
        type: lookupType(astNode),
      };
    }

    return {
      name,
      description,
      deprecationReason,
      fields: Object.fromEntries(
        (astNode.fields ?? []).map((field) => [
          field.name.value,
          buildField(field),
        ]),
      ),
      astNode,
    };
  };

  const buildType = <R extends Role>(
    astNode: TypeDefinitionNode<R>,
  ): IrisTypeDefinition<R> =>
    new IrisTypeDefinition({
      role: astNode.role,
      name: astNode.name.value,
      description: astNode.description?.value,
      variants: () => astNode.variants.map(buildVariant),
      astNode,
    });

  documentAST.definitions.forEach((def) => {
    const name = def.name.value;
    switch (def.kind) {
      case IrisKind.TYPE_DEFINITION:
        typeMap[name] = IrisScalars[name] ?? buildType(def);
        break;
      case IrisKind.DIRECTIVE_DEFINITION:
        directiveDefs.push(def);
        break;
    }
  });

  const assertRoot = (
    operation: string,
    type: IrisTypeDefinition,
  ): IrisTypeDefinition<'resolver'> => {
    if (type && !(type.role === 'resolver' && type.isVariantType)) {
      throw irisNodeError(
        `${operation} root type must be Object type${
          operation === 'Query' ? '' : ' if provided'
        }, it cannot be ${inspect(type)}.`,
        type.astNode,
      );
    }
    return type as IrisTypeDefinition<'resolver'>;
  };

  return new IrisSchema({
    description: undefined,
    types: Object.values(typeMap),
    query: assertRoot('query', typeMap.Query),
    mutation: assertRoot('mutation', typeMap.Mutation),
    subscription: assertRoot('subscription', typeMap.Subscription),
    directives: directiveDefs.map(buildDirective),
  });
}

function getDeprecationReason(node: {
  readonly directives?: ReadonlyArray<DirectiveNode>;
}): IrisMaybe<string> {
  const directiveNode = node.directives?.find(
    (directive) => directive.name.value === GraphQLDeprecatedDirective.name,
  );

  if (directiveNode === undefined) {
    return undefined;
  }

  const reason = directiveNode?.arguments?.find(
    (arg) => arg.name.value === 'reason',
  )?.value;

  return (typeCheckASTValue(reason, IrisScalars.String) as string) ?? '';
}

export type { IrisSchema };
