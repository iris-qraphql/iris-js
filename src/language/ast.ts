import type {
  ArgumentNode,
  BooleanValueNode,
  ConstArgumentNode,
  ConstListValueNode,
  ConstObjectFieldNode,
  ConstObjectValueNode,
  EnumValueNode,
  FloatValueNode,
  IntValueNode,
  Kind,
  ListValueNode,
  Location,
  NullValueNode,
  ObjectFieldNode,
  ObjectValueNode,
  StringValueNode,
  Token,
  VariableNode,
} from 'graphql';

import type { IrisKind } from './kinds';

export { Location, Token };

/**
 * The list of all possible AST node types.
 */
export type ASTNode =
  | NameNode
  | DocumentNode
  | VariableNode
  | ArgumentNode
  | IntValueNode
  | FloatValueNode
  | StringValueNode
  | BooleanValueNode
  | NullValueNode
  | EnumValueNode
  | ListValueNode
  | ObjectValueNode
  | ObjectFieldNode
  | DirectiveNode
  | NamedTypeNode
  | ListTypeNode
  | FieldDefinitionNode
  | ArgumentDefinitionNode
  | ResolverTypeDefinitionNode
  | DataTypeDefinitionNode
  | DirectiveDefinitionNode
  | DataVariantDefinitionNode
  | MaybeTypeNode;

/**
 * @internal
 */
export const QueryDocumentKeys: {
  [NodeT in ASTNode as NodeT['kind']]: ReadonlyArray<keyof NodeT>;
} = {
  Name: [],
  Document: ['definitions'],
  Variable: ['name'],
  Argument: ['name', 'value'],
  IntValue: [],
  FloatValue: [],
  StringValue: [],
  BooleanValue: [],
  NullValue: [],
  EnumValue: [],
  ListValue: ['values'],
  ObjectValue: ['fields'],
  ObjectField: ['name', 'value'],
  Directive: ['name', 'arguments'],
  NamedType: ['name'],
  ListType: ['type'],
  MaybeType: ['type'],
  FieldDefinition: ['description', 'name', 'arguments', 'type', 'directives'],
  InputValueDefinition: [
    'description',
    'name',
    'type',
    'defaultValue',
    'directives',
  ],
  ResolverTypeDefinition: ['description', 'name', 'directives', 'variants'],
  DataTypeDefinition: ['description', 'name', 'directives', 'variants'],
  VariantDefinition: ['name', 'fields'],
  DirectiveDefinition: ['description', 'name', 'arguments', 'locations'],
};

const kindValues = new Set<string>(Object.keys(QueryDocumentKeys));
/**
 * @internal
 */
export function isNode(maybeNode: any): maybeNode is ASTNode {
  const maybeKind = maybeNode?.kind;
  return typeof maybeKind === 'string' && kindValues.has(maybeKind);
}

/** Name */

export interface NameNode {
  readonly kind: Kind.NAME;
  readonly loc?: Location;
  readonly value: string;
}

/** Document */

export interface DocumentNode {
  readonly kind: IrisKind.DOCUMENT;
  readonly loc?: Location;
  readonly definitions: ReadonlyArray<DefinitionNode>;
}

export type DefinitionNode = TypeSystemDefinitionNode;

/** Values */

export type ValueNode =
  | VariableNode
  | IntValueNode
  | FloatValueNode
  | StringValueNode
  | BooleanValueNode
  | NullValueNode
  | EnumValueNode
  | ListValueNode
  | ObjectValueNode;

export type ConstValueNode =
  | IntValueNode
  | FloatValueNode
  | StringValueNode
  | BooleanValueNode
  | NullValueNode
  | EnumValueNode
  | ConstListValueNode
  | ConstObjectValueNode;

export type {
  VariableNode,
  IntValueNode,
  FloatValueNode,
  StringValueNode,
  BooleanValueNode,
  NullValueNode,
  EnumValueNode,
  ListValueNode,
  ConstListValueNode,
  ObjectValueNode,
  ConstObjectValueNode,
  ObjectFieldNode,
  ConstObjectFieldNode,
  ArgumentNode,
  ConstArgumentNode,
};

/** Directives */

export interface DirectiveNode {
  readonly kind: Kind.DIRECTIVE;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentNode>;
}

export interface ConstDirectiveNode {
  readonly kind: Kind.DIRECTIVE;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ConstArgumentNode>;
}

/** Type Reference */

export type WrapperKind = 'LIST' | 'MAYBE';

export type TypeNode = NamedTypeNode | ListTypeNode | MaybeTypeNode;

export interface NamedTypeNode {
  readonly kind: IrisKind.NAMED_TYPE;
  readonly loc?: Location;
  readonly name: NameNode;
}

export interface ListTypeNode {
  readonly kind: IrisKind.LIST_TYPE;
  readonly loc?: Location;
  readonly type: TypeNode;
}

export interface MaybeTypeNode {
  readonly kind: IrisKind.MAYBE_TYPE;
  readonly loc?: Location;
  readonly type: NamedTypeNode | ListTypeNode;
}

/** Type System Definition */

export type TypeSystemDefinitionNode =
  | TypeDefinitionNode
  | DirectiveDefinitionNode;

/** Type Definition */

export type TypeDefinitionNode =
  | ResolverTypeDefinitionNode
  | DataTypeDefinitionNode;

export interface ArgumentDefinitionNode {
  readonly kind: IrisKind.ARGUMENT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly defaultValue?: ConstValueNode;
  readonly directives?: ReadonlyArray<ConstDirectiveNode>;
}

export type Role = 'resolver' | 'data';

export type DataTypeDefinitionNode = TypeDefinition<'data'>;

export type ResolverTypeDefinitionNode = TypeDefinition<'resolver'>;

type ROLE_KIND = {
  resolver: IrisKind.RESOLVER_TYPE_DEFINITION;
  data: IrisKind.DATA_TYPE_DEFINITION;
};

export type TypeDefinition<R extends Role> = {
  readonly kind: ROLE_KIND[R];
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly directives?: ReadonlyArray<ConstDirectiveNode>;
  readonly variants: ReadonlyArray<_VariantDefinitionNode<R>>;
};

export type VariantDefinition<R extends Role> = {
  readonly kind: IrisKind.VARIANT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly directives?: ReadonlyArray<ConstDirectiveNode>;
  readonly name: NameNode;
  readonly fields?: ReadonlyArray<_FieldDefinitionNode<R>>;
};

export type DataFieldDefinitionNode = {
  readonly kind: IrisKind.FIELD_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly directives?: ReadonlyArray<ConstDirectiveNode>;
};

export type FieldDefinitionNode = DataFieldDefinitionNode & {
  readonly arguments?: ReadonlyArray<ArgumentDefinitionNode>;
};

type FIELD_DEF = {
  data: DataFieldDefinitionNode;
  resolver: FieldDefinitionNode;
};

type TYPE_DEF = {
  data: DataTypeDefinitionNode;
  resolver: ResolverTypeDefinitionNode;
};

export type _FieldDefinitionNode<T extends Role> = FIELD_DEF[T];
export type _VariantDefinitionNode<R extends Role> = VariantDefinition<R>;
export type _TypeDefinitionNode<T extends Role> = TYPE_DEF[T];

export type DataVariantDefinitionNode = VariantDefinition<'data'>;
export type ResolverVariantDefinitionNode = VariantDefinition<'resolver'>;

/** Directive Definitions */

export interface DirectiveDefinitionNode {
  readonly kind: IrisKind.DIRECTIVE_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentDefinitionNode>;
  readonly repeatable: boolean;
  readonly locations: ReadonlyArray<NameNode>;
}
