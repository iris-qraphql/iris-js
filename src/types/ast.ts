import type { Kind } from 'graphql';
import { Token } from 'graphql';

import type { Source } from '../parsing/source';
import type { Maybe } from '../utils/type-level';

import { GQLKind, IrisDirectiveLocation, IrisKind } from './kinds';

export { specifiedScalarTypes } from 'graphql';

export type ASTNode =
  | NameNode
  | DocumentNode
  | ArgumentNode
  | StringValueNode
  | NullValueNode
  | DirectiveNode
  | NamedTypeNode
  | ListTypeNode
  | FieldDefinitionNode
  | ArgumentDefinitionNode
  | TypeDefinitionNode
  | VariantDefinitionNode
  | MaybeTypeNode;

export type ValueNode = NullValueNode | StringValueNode;

/**
 * @internal
 */
export const QueryDocumentKeys: {
  [NodeT in ASTNode as NodeT['kind']]: ReadonlyArray<keyof NodeT>;
} = {
  Name: [],
  Document: ['definitions'],
  Argument: ['name', 'value'],
  StringValue: [],
  NullValue: [],
  Directive: ['name', 'arguments'],
  NamedType: ['name'],
  ListType: ['type'],
  MaybeType: ['type'],
  FieldDefinition: ['description', 'name', 'type', 'directives'],
  InputValueDefinition: [
    'description',
    'name',
    'type',
    'defaultValue',
    'directives',
  ],
  TypeDefinition: ['description', 'name', 'directives', 'variants'],
  VariantDefinition: ['description', 'name', 'fields', 'directives'],
};

const kindValues = new Set<string>(Object.keys(QueryDocumentKeys));
/**
 * @internal
 */
export function isNode(maybeNode: any): maybeNode is ASTNode {
  const maybeKind = maybeNode?.kind;
  return typeof maybeKind === 'string' && kindValues.has(maybeKind);
}

export class Location {
  /**
   * The character offset at which this Node begins.
   */
  readonly start: number;

  /**
   * The character offset at which this Node ends.
   */
  readonly end: number;

  /**
   * The Token at which this Node begins.
   */
  readonly startToken: Token;

  /**
   * The Token at which this Node ends.
   */
  readonly endToken: Token;

  /**
   * The Source document the AST represents.
   */
  readonly source: Source;

  constructor(startToken: Token, endToken: Token, source: Source) {
    this.start = startToken.start;
    this.end = endToken.end;
    this.startToken = startToken;
    this.endToken = endToken;
    this.source = source;
  }

  get [Symbol.toStringTag]() {
    return 'Location';
  }

  toJSON(): { start: number; end: number } {
    return { start: this.start, end: this.end };
  }
}

// VALUE
export type StringValueNode = {
  readonly kind: Kind.STRING;
  readonly loc?: Location;
  readonly value: string;
  readonly block?: boolean;
};

export type NullValueNode = {
  readonly kind: Kind.NULL;
  readonly loc?: Location;
};

export type ArgumentNode = {
  readonly kind: Kind.ARGUMENT;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly value: ValueNode;
};

export type DirectiveNode = {
  readonly kind: Kind.DIRECTIVE;
  readonly loc?: Location;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentNode>;
};

/** Name */

export type NameNode = {
  readonly kind: Kind.NAME;
  readonly loc?: Location;
  readonly value: string;
};

/** Document */

export type DocumentNode = {
  readonly kind: IrisKind.DOCUMENT;
  readonly loc?: Location;
  readonly definitions: ReadonlyArray<TypeDefinitionNode>;
};

export { Token };

/** Type Reference */

export type WrapperKind = 'LIST' | 'MAYBE' | 'NAMED';

export type TypeNode = NamedTypeNode | ListTypeNode | MaybeTypeNode;

export type NamedTypeNode = {
  readonly kind: IrisKind.NAMED_TYPE;
  readonly loc?: Location;
  readonly name: NameNode;
};

export type ListTypeNode = {
  readonly kind: IrisKind.LIST_TYPE;
  readonly loc?: Location;
  readonly type: TypeNode;
};

export type MaybeTypeNode = {
  readonly kind: IrisKind.MAYBE_TYPE;
  readonly loc?: Location;
  readonly type: NamedTypeNode | ListTypeNode;
};

export type ArgumentDefinitionNode = {
  readonly kind: IrisKind.ARGUMENT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly defaultValue?: ValueNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
};

export type FieldDefinitionNode = {
  readonly kind: IrisKind.FIELD_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly type: TypeNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
};

export type VariantDefinitionNode = {
  readonly kind: IrisKind.VARIANT_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
  readonly isTypeVariantNode: boolean;
  readonly name: NameNode;
  readonly fields?: ReadonlyArray<FieldDefinitionNode>;
};

export type TypeDefinitionNode = {
  readonly kind: IrisKind.TYPE_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly directives?: ReadonlyArray<DirectiveNode>;
  readonly variants: ReadonlyArray<VariantDefinitionNode>;
};

export interface DirectiveDefinitionNode {
  readonly kind: IrisKind.DIRECTIVE_DEFINITION;
  readonly loc?: Location;
  readonly description?: StringValueNode;
  readonly name: NameNode;
  readonly arguments?: ReadonlyArray<ArgumentDefinitionNode>;
  readonly repeatable: boolean;
  readonly locations: ReadonlyArray<NameNode>;
}

export const isTypeDefinitionNode = (
  node: ASTNode,
): node is TypeDefinitionNode => node.kind === IrisKind.TYPE_DEFINITION;

export const isTypeVariantNode = ({ name, variants }: TypeDefinitionNode) => {
  const typeName = name.value;
  const [variant] = variants;

  if (variants.length === 0) {
    return true;
  }

  return (
    variants.length === 1 &&
    variant.name.value === typeName &&
    variant.fields !== undefined
  );
};

export const getRefTypeName = (node: TypeNode): NameNode => {
  switch (node.kind) {
    case IrisKind.NAMED_TYPE:
      return node.name;
    case IrisKind.LIST_TYPE:
      return getRefTypeName(node.type);
    case IrisKind.MAYBE_TYPE:
      return getRefTypeName(node.type);
  }
};

export const getVariant = (
  { variants }: TypeDefinitionNode,
  name?: string,
): Maybe<VariantDefinitionNode> =>
  name ? variants.find((v) => v.name.value === name) : variants[0];

export const isRequiredArgument = (arg: ArgumentDefinitionNode) =>
  arg.type.kind !== IrisKind.MAYBE_TYPE && arg.defaultValue === undefined;

export const getDeprecationReason = (
  node: ASTNode & Pick<TypeDefinitionNode, 'directives'>,
): Maybe<string> => {
  const deprecated = node.directives?.find(
    (d) => d.name.value === 'deprecated',
  );

  if (!deprecated) {
    return undefined;
  }

  const reason = deprecated.arguments?.[0]?.value as Maybe<StringValueNode>;
  return reason?.value ?? '';
};

type IrisDirective = {
  name: string;
  description?: string;
  locations: ReadonlyArray<IrisDirectiveLocation>;
  args: ReadonlyArray<ArgumentDefinitionNode>;
  astNode?: DirectiveDefinitionNode;
};

export const specifiedDirectives: ReadonlyArray<IrisDirective> = Object.freeze([
  {
    name: 'deprecated',
    description: 'Marks an element of a GraphQL schema as no longer supported.',
    locations: [
      IrisDirectiveLocation.FIELD_DEFINITION,
      IrisDirectiveLocation.ARGUMENT_DEFINITION,
      IrisDirectiveLocation.VARIANT_DEFINITION,
    ],
    args: [
      {
        kind: IrisKind.ARGUMENT_DEFINITION,
        name: { value: 'reason', kind: GQLKind.NAME },
        description: {
          kind: GQLKind.STRING,
          value:
            'Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).',
        },
        defaultValue: { value: '', kind: GQLKind.STRING },
        type: {
          kind: IrisKind.NAMED_TYPE,
          name: { value: 'string', kind: GQLKind.NAME },
        },
      },
    ],
  },
]);
