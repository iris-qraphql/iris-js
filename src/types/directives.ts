import { assertName } from 'graphql';

import { inspect, instanceOf } from '../utils/legacy';
import type { Maybe } from '../utils/type-level';

import type { ArgumentDefinitionNode, DirectiveDefinitionNode } from './ast';
import { IrisDirectiveLocation } from './directiveLocation';

/**
 * Test if the given value is a GraphQL directive.
 */
export function isDirective(directive: unknown): directive is GraphQLDirective {
  return instanceOf(directive, GraphQLDirective);
}

export function assertDirective(directive: unknown): GraphQLDirective {
  if (!isDirective(directive)) {
    throw new Error(
      `Expected ${inspect(directive)} to be a GraphQL directive.`,
    );
  }
  return directive;
}

/**
 * Directives are used by the GraphQL runtime as a way of modifying execution
 * behavior. Type system creators will usually not create these directly.
 */
export class GraphQLDirective {
  name: string;
  description: Maybe<string>;
  locations: ReadonlyArray<IrisDirectiveLocation>;
  // args: ReadonlyArray<IrisArgument>;
  isRepeatable: boolean;
  astNode: Maybe<DirectiveDefinitionNode>;

  constructor(config: Readonly<GraphQLDirectiveConfig>) {
    this.name = assertName(config.name);
    this.description = config.description;
    this.locations = config.locations;
    this.isRepeatable = config.isRepeatable ?? false;
    this.astNode = config.astNode;
    // this.args = Object.values(config.args ?? {});
  }

  get [Symbol.toStringTag]() {
    return 'GraphQLDirective';
  }

  toString(): string {
    return '@' + this.name;
  }

  toJSON(): string {
    return this.toString();
  }
}

export interface GraphQLDirectiveConfig {
  name: string;
  description?: Maybe<string>;
  locations: ReadonlyArray<IrisDirectiveLocation>;
  args?: ReadonlyArray<ArgumentDefinitionNode>;
  isRepeatable?: Maybe<boolean>;
  astNode?: Maybe<DirectiveDefinitionNode>;
}

/**
 * Used to declare element of a GraphQL schema as deprecated.
 */
export const GraphQLDeprecatedDirective: GraphQLDirective =
  new GraphQLDirective({
    name: 'deprecated',
    description: 'Marks an element of a GraphQL schema as no longer supported.',
    locations: [
      IrisDirectiveLocation.FIELD_DEFINITION,
      IrisDirectiveLocation.ARGUMENT_DEFINITION,
    ],
    // args: [
    //   {
    //     name: 'reason',
    //     type: { },
    //     description: 'Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).',
    //     defaultValue: '',
    //   },
    // ],
  });

/**
 * The full list of specified directives.
 */
export const specifiedDirectives: ReadonlyArray<GraphQLDirective> =
  Object.freeze([GraphQLDeprecatedDirective]);

export function isSpecifiedDirective(directive: GraphQLDirective): boolean {
  return specifiedDirectives.some(({ name }) => name === directive.name);
}
