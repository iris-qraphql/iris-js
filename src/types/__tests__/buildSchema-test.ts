import { print } from '../../printing/printer';
import { dedent } from '../../utils/dedent';

import { getDeprecationReason, getVariant } from '../ast';
import { GraphQLDeprecatedDirective } from '../directives';
import { buildSchema, getDirective, getType } from '../schema';

const cycleSDL = (sdl: string): string => print(buildSchema(sdl));

describe('Schema Builder', () => {
  it('Empty type', () => {
    const sdl = dedent`
      data EmptyType
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Simple type', () => {
    const sdl = dedent`
      data Type = Type {
        str: String
        int: Int
        float: Float
        id: ID
        bool: Boolean
      }
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('include standard type only if it is used', () => {
    const schema = buildSchema('data Query');

    // String and Boolean are always included through introspection types
    expect(getType(schema, 'Int')).toEqual(undefined);
    expect(getType(schema, 'Float')).toEqual(undefined);
    expect(getType(schema, 'ID')).toEqual(undefined);
  });

  it('With directives', () => {
    const sdl = dedent`
      directive @foo(arg: Int) on FIELD

      directive @repeatableFoo(arg: Int) repeatable on FIELD
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Supports descriptions', () => {
    const sdl = dedent`
      """This is a directive"""
      directive @foo(
        """It has an argument"""
        arg: Int
      ) on FIELD

      """This is a data type"""
      data Foo = Foo {
        """It has a field"""
        field: Int
      }

      """There is nothing inside!"""
      data BlackHole

      """With an Enum"""
      data Color = RED {} | GREEN {} | BLUE {}
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Overriding directives excludes specified', () => {
    const schema = buildSchema(`
      directive @skip on FIELD
      directive @include on FIELD
      directive @deprecated on FIELD_DEFINITION
      directive @specifiedBy on FIELD_DEFINITION
    `);

    expect(Object.values(schema.directives)).toHaveLength(4);
    expect(getDirective(schema, 'deprecated')).not.toEqual(
      GraphQLDeprecatedDirective,
    );
  });

  it('Type modifiers', () => {
    const sdl = dedent`
      data Type = Type {
        nonNullStr: String?
        listOfStrings: [String]
        listOfNonNullStrings: [String?]
        nonNullListOfStrings: [String]?
        nonNullListOfNonNullStrings: [String?]?
      }
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Recursive type', () => {
    const sdl = dedent`
      data Type = Type {
        str: String
        recurse: Type
      }
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Two types circular', () => {
    const sdl = dedent`
      data One = One {
        str: String
        typeTwo: Two?
      }

      data Two = Two {
        str: String?
        typeOne: One
      }
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Empty Enum', () => {
    const sdl = dedent`
      data EmptyEnum
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Simple output Enum', () => {
    const sdl = dedent`
      data Hello = WORLD {}
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Multiple value Enum', () => {
    const sdl = dedent`
      data Hello = WO {} | RLD {}
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Empty union', () => {
    const sdl = dedent`
      data EmptyUnion
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Simple Union', () => {
    const sdl = dedent`
      data World = {
        str: String
      }

      data Hello = World
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Multiple Union', () => {
    const sdl = dedent`
      data Hello = WorldOne | WorldTwo

      data WorldOne = {
        str: String
      }

      data WorldTwo = {
        str: String
      }
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Empty Input Object', () => {
    const sdl = dedent`
      data EmptyInputObject
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Simple Input Object', () => {
    const sdl = dedent`
      data Type = Type {
        str: String
        int: Int
        bool: Boolean
      }
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Unreferenced type implementing referenced union', () => {
    const sdl = dedent`
      data Concrete = {
        key: String
      }

      data Query = {
        union: Union
      }

      data Union = Concrete
    `;
    expect(cycleSDL(sdl)).toEqual(sdl);
  });

  it('Supports @deprecated', () => {
    const sdl = dedent`
      data MyEnum = VALUE {} | OLD_VALUE @deprecated {} | OTHER_VALUE @deprecated(reason: "Terrible reasons") {}

      data MyInput = {
        oldInput: String? @deprecated
        otherInput: String @deprecated(reason: "Use newInput")
        newInput: String
      }

      data Type = {
        field1: String @deprecated
        field2: Int @deprecated(reason: "Because I said so")
        enum: MyEnum
        field5: String
      }
    `;

    expect(cycleSDL(sdl)).toEqual(sdl);

    const schema = buildSchema(sdl);

    expect(
      getType(schema, 'MyEnum')?.variants.map((variant) => ({
        name: variant.name.value,
        deprecation: getDeprecationReason(variant),
      })),
    ).toMatchSnapshot();

    const rootType = getType(schema, 'Type');
    const variant = rootType ? getVariant(rootType) : undefined;

    expect(
      variant?.fields?.map((f) => ({
        name: f.name.value,
        deprecation: getDeprecationReason(f),
      })),
    ).toMatchSnapshot();
  });

  it('Rejects invalid SDL', () => {
    const sdl = `
      data Query = {
        foo: String @unknown
      }
    `;
    expect(() => buildSchema(sdl)).toThrow('Unknown directive "@unknown".');
  });
});
