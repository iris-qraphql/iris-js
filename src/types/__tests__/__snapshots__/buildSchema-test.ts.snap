// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Builder Supports @deprecated 1`] = `
Array [
  Object {
    "deprecation": undefined,
    "name": Object {
      "kind": "Name",
      "loc": Object {
        "end": 19,
        "start": 14,
      },
      "value": "VALUE",
    },
  },
  Object {
    "deprecation": undefined,
    "name": Object {
      "kind": "Name",
      "loc": Object {
        "end": 34,
        "start": 25,
      },
      "value": "OLD_VALUE",
    },
  },
  Object {
    "deprecation": undefined,
    "name": Object {
      "kind": "Name",
      "loc": Object {
        "end": 63,
        "start": 52,
      },
      "value": "OTHER_VALUE",
    },
  },
]
`;
